File: apps/cli/src/commands/kv.rs
use crate::{
    api_types::{
        CreateStoreRequest, CreateStoreResponse, GetValueResponse, KvStore, ListKeysResponse,
        SetValueRequest,
    },
    auth::AuthRequestBuilder,
    cli::KvCommands,
    config::CliConfig,
};
use color_eyre::eyre::{Result, eyre};
use reqwest::Client;

pub(crate) async fn handle_kv_command(
    client: &Client,
    config: &CliConfig,
    cmd: KvCommands,
) -> Result<()> {
    match cmd {
        KvCommands::CreateStore { guild, name } => {
            let url = format!("{}/kv/stores", config.api_url);
            let body = CreateStoreRequest {
                guild_id: guild.clone(),
                store_name: name.clone(),
            };
            let resp: CreateStoreResponse = client
                .post(url)
                .maybe_bearer(&config.token)?
                .json(&body)
                .send()
                .await?
                .error_for_status()?
                .json()
                .await?;
            println!(
                "Created KV store '{}' for guild {}",
                resp.store.store_name, resp.store.guild_id
            );
        }
        KvCommands::ListStores { guild } => {
            let url = format!("{}/kv/stores?guild_id={}", config.api_url, guild);
            let stores: Vec<KvStore> = client
                .get(url)
                .maybe_bearer(&config.token)?
                .send()
                .await?
                .error_for_status()?
                .json()
                .await?;
            if stores.is_empty() {
                println!("No KV stores found for guild {}", guild);
            } else {
                println!("KV stores for guild {}:", guild);
                for store in stores {
                    println!("  - {}", store.store_name);
                }
            }
        }
        KvCommands::DeleteStore { guild, name } => {
            let url = format!("{}/kv/stores/{}/{}", config.api_url, guild, name);
            client
                .delete(url)
                .maybe_bearer(&config.token)?
                .send()
                .await?
                .error_for_status()?;
            println!("Deleted KV store '{}' for guild {}", name, guild);
        }
        KvCommands::Set {
            guild,
            store,
            key,
            value,
            expiration,
            metadata,
        } => {
            let url = format!("{}/kv/{}/{}/{}", config.api_url, guild, store, key);
            let metadata_json: Option<serde_json::Value> = match metadata {
                Some(ref m) => Some(
                    serde_json::from_str(m).map_err(|e| eyre!("Invalid JSON metadata: {}", e))?,
                ),
                None => None,
            };
            let body = SetValueRequest {
                value: value.clone(),
                expiration,
                metadata: metadata_json,
            };
            client
                .put(url)
                .maybe_bearer(&config.token)?
                .json(&body)
                .send()
                .await?
                .error_for_status()?;
            let mut msg = format!(
                "Set {}={} in store '{}' for guild {}",
                key, value, store, guild
            );
            if let Some(exp) = expiration {
                msg.push_str(&format!(" (expires: {})", exp));
            }
            println!("{}", msg);
        }
        KvCommands::Get { guild, store, key } => {
            let url = format!("{}/kv/{}/{}/{}", config.api_url, guild, store, key);
            let resp: GetValueResponse = client
                .get(url)
                .maybe_bearer(&config.token)?
                .send()
                .await?
                .error_for_status()?
                .json()
                .await?;
            match resp.value {
                Some(value) => println!("{}", value),
                None => println!("Key '{}' not found", key),
            }
        }
        KvCommands::Delete { guild, store, key } => {
            let url = format!("{}/kv/{}/{}/{}", config.api_url, guild, store, key);
            client
                .delete(url)
                .maybe_bearer(&config.token)?
                .send()
                .await?
                .error_for_status()?;
            println!(
                "Deleted key '{}' from store '{}' for guild {}",
                key, store, guild
            );
        }
        KvCommands::ListKeys {
            guild,
            store,
            prefix,
            limit,
        } => {
            let mut url = format!("{}/kv/{}/{}", config.api_url, guild, store);
            let mut params = Vec::new();
            if let Some(p) = prefix {
                params.push(format!("prefix={}", p));
            }
            if let Some(l) = limit {
                params.push(format!("limit={}", l));
            }
            if !params.is_empty() {
                url = format!("{}?{}", url, params.join("&"));
            }
            let resp: ListKeysResponse = client
                .get(url)
                .maybe_bearer(&config.token)?
                .send()
                .await?
                .error_for_status()?
                .json()
                .await?;
            if resp.keys.is_empty() {
                println!("No keys found in store '{}'", store);
            } else {
                println!(
                    "Keys in store '{}' ({} of {} shown):",
                    store,
                    resp.keys.len(),
                    if resp.list_complete { "all" } else { "partial" }
                );
                for key in &resp.keys {
                    let mut line = format!("  - {}", key.name);
                    if let Some(exp) = key.expiration {
                        line.push_str(&format!(" (expires: {})", exp));
                    }
                    if let Some(ref meta) = key.metadata {
                        line.push_str(&format!(" [metadata: {}]", meta));
                    }
                    println!("{}", line);
                }
                if !resp.list_complete {
                    println!(
                        "\nMore keys available. Use --limit {} to get more, or --cursor {} for next page.",
                        limit.unwrap_or(100),
                        resp.cursor.clone().unwrap_or_default()
                    );
                }
            }
        }
    }
    Ok(())
}

File: apps/cli/src/commands/kv.ts
import type { components } from '../generated/openapi-schema'
import { authHeaders, createApiClient, expectOk } from '../lib/http'
import { promptIfMissing } from '../lib/prompts'
import type { CliConfig } from '../lib/types'

export async function createStore(
  config: CliConfig,
  guildArg?: string,
  nameArg?: string
): Promise<void> {
  const guild = await promptIfMissing(guildArg, 'Guild ID')
  const name = await promptIfMissing(nameArg, 'Store name')

  const client = createApiClient(config)
  const response = await expectOk<components['schemas']['CreateStoreResponse']>(
    client.POST('/kv/api/kv/stores', {
      headers: authHeaders(config),
      body: {
        guild_id: guild,
        store_name: name
      }
    })
  )

  process.stdout.write(
    `Created KV store '${response.store.store_name}' for guild ${response.store.guild_id}\n`
  )
}

export async function listStores(config: CliConfig, guildArg?: string): Promise<void> {
  const guild = await promptIfMissing(guildArg, 'Guild ID')

  const client = createApiClient(config)
  const stores = await expectOk<components['schemas']['KvStore'][]>(
    client.GET('/kv/api/kv/stores', {
      headers: authHeaders(config),
      params: { query: { guild_id: guild } }
    })
  )

  if (stores.length === 0) {
    process.stdout.write(`No KV stores found for guild ${guild}\n`)
    return
  }

  process.stdout.write(`KV stores for guild ${guild}:\n`)
  for (const store of stores) {
    process.stdout.write(`  - ${store.store_name}\n`)
  }
}

export async function deleteStore(
  config: CliConfig,
  guildArg?: string,
  nameArg?: string
): Promise<void> {
  const guild = await promptIfMissing(guildArg, 'Guild ID')
  const name = await promptIfMissing(nameArg, 'Store name')

  const client = createApiClient(config)
  await expectOk(
    client.DELETE('/kv/api/kv/stores/{guild_id}/{store_name}', {
      headers: authHeaders(config),
      params: {
        path: {
          guild_id: guild,
          store_name: name
        }
      }
    })
  )

  process.stdout.write(`Deleted KV store '${name}' for guild ${guild}\n`)
}

export async function setValue(
  config: CliConfig,
  guildArg: string | undefined,
  storeArg: string | undefined,
  keyArg: string | undefined,
  valueArg: string | undefined,
  expiration?: number,
  metadata?: string
): Promise<void> {
  const guild = await promptIfMissing(guildArg, 'Guild ID')
  const store = await promptIfMissing(storeArg, 'Store name')
  const key = await promptIfMissing(keyArg, 'Key')
  const value = await promptIfMissing(valueArg, 'Value')

  const metadataValue = metadata ? JSON.parse(metadata) : undefined

  const client = createApiClient(config)
  await expectOk(
    client.PUT('/kv/api/kv/{guild_id}/{store_name}/{key}', {
      headers: authHeaders(config),
      params: {
        path: {
          guild_id: guild,
          store_name: store,
          key
        }
      },
      body: {
        value,
        expiration,
        metadata: metadataValue
      }
    })
  )

  process.stdout.write(`Set ${key}=${value} in store '${store}' for guild ${guild}\n`)
}

export async function getValue(
  config: CliConfig,
  guildArg?: string,
  storeArg?: string,
  keyArg?: string
): Promise<void> {
  const guild = await promptIfMissing(guildArg, 'Guild ID')
  const store = await promptIfMissing(storeArg, 'Store name')
  const key = await promptIfMissing(keyArg, 'Key')

  const client = createApiClient(config)
  const response = await expectOk<components['schemas']['GetValueResponse']>(
    client.GET('/kv/api/kv/{guild_id}/{store_name}/{key}', {
      headers: authHeaders(config),
      params: {
        path: {
          guild_id: guild,
          store_name: store,
          key
        }
      }
    })
  )

  if (response.value == null) {
    process.stdout.write(`Key '${key}' not found\n`)
    return
  }

  process.stdout.write(`${response.value}\n`)
}

export async function deleteValue(
  config: CliConfig,
  guildArg?: string,
  storeArg?: string,
  keyArg?: string
): Promise<void> {
  const guild = await promptIfMissing(guildArg, 'Guild ID')
  const store = await promptIfMissing(storeArg, 'Store name')
  const key = await promptIfMissing(keyArg, 'Key')

  const client = createApiClient(config)
  await expectOk(
    client.DELETE('/kv/api/kv/{guild_id}/{store_name}/{key}', {
      headers: authHeaders(config),
      params: {
        path: {
          guild_id: guild,
          store_name: store,
          key
        }
      }
    })
  )

  process.stdout.write(`Deleted key '${key}' from store '${store}' for guild ${guild}\n`)
}

export async function listKeys(
  config: CliConfig,
  guildArg?: string,
  storeArg?: string,
  prefix?: string,
  limit?: number,
  cursor?: string
): Promise<void> {
  const guild = await promptIfMissing(guildArg, 'Guild ID')
  const store = await promptIfMissing(storeArg, 'Store name')

  const client = createApiClient(config)
  const response = await expectOk(
    client.GET('/kv/api/kv/{guild_id}/{store_name}', {
      headers: authHeaders(config),
      params: {
        path: {
          guild_id: guild,
          store_name: store
        },
        query: {
          prefix,
          limit,
          cursor
        }
      }
    })
  )

  if (response.keys.length === 0) {
    process.stdout.write(`No keys found in store '${store}'\n`)
    return
  }

  process.stdout.write(`Keys in store '${store}' (${response.keys.length} shown):\n`)
  for (const key of response.keys) {
    const expires = key.expiration ? ` (expires: ${key.expiration})` : ''
    const meta = key.metadata ? ` [metadata: ${JSON.stringify(key.metadata)}]` : ''
    process.stdout.write(`  - ${key.name}${expires}${meta}\n`)
  }

  const listComplete = 'list_complete' in response ? response.list_complete : response.listComplete
  if (!listComplete && response.cursor) {
    process.stdout.write(`More keys available. Use --cursor ${response.cursor}\n`)
  }
}
